### PR TITLE
jwt-auth: changed the interface of the class for JWT verification

### PR DESCRIPTION
**Pull request by mtakigiku**
as commented in the [former PR](https://github.com/istio/proxy/pull/410#issuecomment-323219386)

### CODE REVIEW

File: src/envoy/auth/jwt.cc [hunk 1]
```diff
@@ -208,7 +196,38 @@ class EvpPkeyGetter {
   }
 };
 
-// Class to decode and verify JWT. Setup() must be called before
+std::pair<bool, std::string> GetStringFromJson(const rapidjson::Document &d,
```
Comment (lizan) [inline src/envoy/auth/jwt.cc]: How about reusing envoy's JsonLoader and JsonObject?
https://github.com/lyft/envoy/blob/master/include/envoy/json/json_object.h

File: src/envoy/auth/jwt_test.cc [hunk 2]
```diff
@@ -307,11 +307,23 @@ class JwtTestWithJwk : public testing::Test {
       "37BD2KXH3jN9_EBocxwcxhkPwSLtP3dgkfls_f5GoWCgmp-c5ycIskCDcIjxRnPjg";
 };
 
+#define TEST_SETUP(public_key, jwt, verified) \
```
Comment (lizan) [inline src/envoy/auth/jwt_test.cc]: Use "Value Parameterized Tests".

See https://github.com/google/googletest/blob/master/googletest/docs/AdvancedGuide.md

File: src/envoy/auth/jwt.cc (hunk 1 above)
Comment (mtakigiku) [reply src/envoy/auth/jwt.cc]: you mean, replacing all `rapidjson::Document` in this lib with envoy's JsonObject? Or just for this function?

Comment (lizan) [reply src/envoy/auth/jwt.cc]: yes replacing rapidjson::Document if possible.

File: src/envoy/auth/jwt_test.cc (hunk 2 above)
Comment (mtakigiku) [reply src/envoy/auth/jwt_test.cc]: Do you mean something like this?:
```
vector<tuple<string, string, bool> > testdata = {
  make_tuple(jwt1, pubkey1, verify_result1),
  make_tuple(jwt2, pubkey2, verify_result2),
  ...
};
INSTANTIATE_TEST_CASE_P(foobar, JwtTest, testing::ValuesIn(testdata));
...
```

File: src/envoy/auth/jwt.cc (hunk 1 above)
Comment (mtakigiku) [reply src/envoy/auth/jwt.cc]: done

File: src/envoy/auth/jwt_test.cc (hunk 2 above)
Comment (mtakigiku) [reply src/envoy/auth/jwt_test.cc]: done

File: src/envoy/auth/jwt_test.cc [hunk 3]
```diff
@@ -307,242 +316,206 @@ class JwtTestWithJwk : public testing::Test {
...
 }
 
-TEST_F(JwtTestWithJwk, JwkBadPublicKey) {
-  std::string invalid_pubkey = R"EOF({"keys":[]})EOF";
-  JwtVerifier v;
-  auto payload = v.Decode(*Pubkeys::ParseFromJwks(invalid_pubkey), kJwtNoKid);
-  EXPECT_FALSE(payload);
-  EXPECT_EQ(v.GetStatus(), Status::JWK_NO_VALID_PUBKEY);
-}
+INSTANTIATE_TEST_CASE_P(TestCases, JwtTest, testing::ValuesIn(TestData()));
```
Comment (qiwzhang) [inline src/envoy/auth/jwt_test.cc]: I am worrying about if one of tests fail,  how easy to identify which one file,  does each test has a name to associate with it?

File: src/envoy/auth/jwt.h [hunk 4]
```diff
@@ -134,12 +183,12 @@ class JwtVerifier : public WithStatus {
 class Pubkeys : public WithStatus {
  public:
   Pubkeys(){};
-  static std::unique_ptr<Pubkeys> ParseFromPem(const std::string &pkey_pem);
-  static std::unique_ptr<Pubkeys> ParseFromJwks(const std::string &pkey_jwks);
+  static std::unique_ptr<Pubkeys> ParseFromPem(const std::string& pkey_pem);
```
[the remark this answers is not in the document]
Comment (mtakigiku) [reply src/envoy/auth/jwt.h]: done

File: src/envoy/auth/jwt_test.cc (hunk 3 above)
Comment (mtakigiku) [reply src/envoy/auth/jwt_test.cc]: No, each test has only a number (corresponding to what-th test it is), and I agree that it is not so easy to see if it fails...

File: src/envoy/auth/jwt.cc [hunk 5]
```diff
@@ -246,22 +228,48 @@ class Verifier {
...
+    // Parse header json
+    header_str_base64url_ = jwt_split[0];
+    header_str_ = Base64UrlDecode(jwt_split[0]);
+    try {
+      header_ = Json::Factory::loadFromString(header_str_);
+    } catch (...) {
       UpdateStatus(Status::JWT_HEADER_PARSE_ERROR);
       return false;
-    }
+    };
```
Comment (lizan) [inline src/envoy/auth/jwt.cc]: ?

File: src/envoy/auth/jwt_test.cc (hunk 3 above)
Comment (lizan) [reply src/envoy/auth/jwt_test.cc]: IMO it is fine with numbers.

Comment (mtakigiku) [reply src/envoy/auth/jwt_test.cc]: I fixed it anyway

File: src/envoy/auth/jwt.cc [hunk 6]
```diff
@@ -344,7 +355,81 @@ class Verifier {
   }
 };
 
-}  // namespace
+JwtVerifier::JwtVerifier(const std::string &jwt) {
+  impl_ = new Impl();
```
[the remark this answers is not in the document]
Comment (mtakigiku) [reply src/envoy/auth/jwt.cc]: Got a compile error if I use unique_ptr:
```
/usr/include/c++/5/bits/unique_ptr.h:74:22: error: invalid application of 'sizeof' to incomplete type 'Envoy::Http::Auth::JwtVerifier::Impl'
```
How can we fix it?

**Pull request merged**